File: interdisjv/src/main/java/org/daydevjv/interdisjv/ch12/Exercises12.java
package org.daydevjv.interdisjv.ch12;

import java.util.Random;

public class Exercises12 {

    @SuppressWarnings({"UnnecessaryLocalVariable", "UnusedAssignment"})
    private static void ex1_2_1() {
        System.out.println();
        int a = 10;
        int b = 20;
        int t = a;
        b = t;
        a = b;
        System.out.println("a = " + a);
        System.out.println("b = " + b);
    }

    private static void ex1_2_2() {
        System.out.println();
        System.out.println("r = " + sumOfSquares(0.25));
        System.out.println("r = " + sumOfSquares(0.5));
        System.out.println("r = " + sumOfSquares(0.75));
    }

    private static double sumOfSquares(double a) {
        double cos = Math.cos(a);
        double sin = Math.sin(a);
        return cos * cos + sin * sin;
    }

    @SuppressWarnings("ConstantConditions")
    private static void ex1_2_3() {
        System.out.println();
        System.out.println(longBooleanExpression(true, true));
        System.out.println(longBooleanExpression(false, true));
        System.out.println(longBooleanExpression(true, false));
        System.out.println(longBooleanExpression(false, false));
    }

    @SuppressWarnings("ConstantConditions")
    private static boolean longBooleanExpression(boolean a, boolean b) {
        return (!(a && b) && (a || b)) || ((a && b) || !(a || b));
    }

    // ex1_2_11 in py
    private static void ex1_2_14() {
        System.out.println();
        evenlyDivide(8, 2);
        evenlyDivide(3, 9);
        evenlyDivide(4, 11);
    }

    private static void evenlyDivide(int a, int b) {
        boolean r = a % b == 0 || b % a == 0;
        System.out.println(r);
    }

    // ex1_2_12 in py
    private static void ex1_2_15() {
        System.out.println();
        System.out.println(canMakeTriangle(3, 4, 5));
        System.out.println(canMakeTriangle(2, 4, 7));
    }

    private static boolean canMakeTriangle(int a, int b, int c) {
        boolean oneIsBigger = a >= b + c || b >= a + c || c >= b + a;
        return !oneIsBigger;
    }

    // ex1_2_16 in py
    private static void ex1_2_19() {
        System.out.println();
        System.out.println(randomBetween(0, 10));
        System.out.println(randomBetween(3, 4));
        System.out.println(randomBetween(-1, 1));
    }

    private static int randomBetween(int from, int to) {
        Random random = new Random();
        return from + random.nextInt(to - from + 1);
    }

    // ex1_2_17 in py
    private static void ex1_2_20() {
        System.out.println();
        Random random = new Random();
        int a = random.nextInt(5) + 1;
        int b = random.nextInt(5) + 1;
        System.out.println(a + b);
    }

    // ex1_2_20 in py
    private static void ex1_2_23() {
        System.out.println();
        System.out.println(withinInterval(19, 3));
        System.out.println(withinInterval(20, 3));
        System.out.println(withinInterval(21, 3));
        System.out.println(withinInterval(15, 5));
        System.out.println(withinInterval(20, 6));
        System.out.println(withinInterval(19, 6));
        System.out.println(withinInterval(21, 6));
    }

    private static boolean withinInterval(int day, int month) {
        final int startDay = 20;
        final int startMonth = 3;
        final int endDay = 20;
        final int endMonth = 6;
        boolean out = false;
        if (month < startMonth || month > endMonth) {
            out = true;
        } else if (month == startMonth) {
            out = day < startDay;
        } else if (month == endMonth) {
            out = day > endDay;
        }
        return !out;
    }

    private static void ex1_2_24() {
        System.out.println();
        System.out.println(continuouslyCompoundedInterest(10_000, 3, 0.05));
        System.out.println(continuouslyCompoundedInterest(20_000, 10, 0.15));
    }

    private static double continuouslyCompoundedInterest(double investment, int years, double annualInterestRate) {
        return investment * Math.pow(Math.E, annualInterestRate * years);
    }

    // ex1_2_31 in py
    private static void ex1_2_34() {
        System.out.println();
        threeSort(10, 30, 20);
        threeSort(30, 20, 10);
        threeSort(20, 10, 30);
    }

    private static void threeSort(int a, int b, int c) {
        int tmp;
        if (b < a) {
            tmp = a;
            a = b;
            b = tmp;
        }
        if (c < a) {
            tmp = a;
            a = c;
            c = tmp;
        }
        if (c < b) {
            tmp = b;
            b = c;
            c = tmp;
        }
        System.out.println(a + " " + b + " " + c);
    }

    public static void main(String[] args) {
        ex1_2_1();
        ex1_2_2();
        ex1_2_3();
        ex1_2_14();
        ex1_2_15();
        ex1_2_19();
        ex1_2_20();
        ex1_2_23();
        ex1_2_24();
        ex1_2_34();
    }
}

File: interdispy/ch12/exercises12.py
import math
import random


# noinspection PyUnusedLocal
def ex1_2_1():
    print()
    a = 10
    b = 20
    t = a
    b = t
    a = b
    print('a = ', a)
    print('b = ', b)


def sum_of_squares(a):
    sin = math.sin(a)
    cos = math.cos(a)
    return sin * sin + cos * cos


def ex1_2_2():
    print()
    print("r = ", sum_of_squares(0.25))
    print("r = ", sum_of_squares(0.5))
    print("r = ", sum_of_squares(0.75))


def long_boolean_expression(a, b):
    return (not(a and b) and (a or b)) or ((a and b) or not(a or b))


def ex1_2_3():
    print()
    print(long_boolean_expression(True, True))
    print(long_boolean_expression(False, True))
    print(long_boolean_expression(True, False))
    print(long_boolean_expression(False, False))


# ex1_2_14 in java
def ex1_2_11():
    print()
    evenly_divide(8, 2)
    evenly_divide(2, 8)
    evenly_divide(4, 11)


def evenly_divide(a, b):
    r = a % b == 0 or b % a == 0
    print(r)


# ex1_2_15 in java
def ex1_2_12():
    print()
    print(can_make_triangle(3, 4, 5))
    print(can_make_triangle(2, 4, 7))


def can_make_triangle(a, b, c):
    bigger = a >= b + c or b >= a + c or c >= a + b
    return not bigger


# ex1_2_19 in java
def ex1_2_16():
    print()
    print(random.randint(0, 10))
    print(random.randint(3, 4))
    print(random.randint(-1, 1))


# ex1_2_20 in java
def ex1_2_17():
    print()
    a = random.randint(1, 6)
    b = random.randint(1, 6)
    print(a + b)


# ex1_2_23 in java
def ex1_2_20():
    print()
    print(within_interval(19, 3))
    print(within_interval(20, 3))
    print(within_interval(21, 3))
    print(within_interval(15, 5))
    print(within_interval(20, 6))
    print(within_interval(19, 6))
    print(within_interval(21, 6))


def within_interval(day, month):
    start_day = 20
    start_month = 3
    end_day = 20
    end_month = 6
    out = False
    if month < start_month or end_month < month:
        out = True
    elif month == start_month:
        out = day < start_day
    elif month == end_month:
        out = day > end_day
    return not out


def ex1_2_21():
    print()
    print(continuously_compounded_interest(10000, 3, 0.05))
    print(continuously_compounded_interest(20000, 10, 0.15))


def continuously_compounded_interest(investment, years, annual_rate):
    return investment * (math.e ** (annual_rate * years))


# ex1_2_34 in java
def ex1_2_31():
    print()
    three_sort(10, 30, 20)
    three_sort(30, 20, 10)
    three_sort(20, 10, 30)


def three_sort(a, b, c):
    if b < a:
        t = a
        a = b
        b = t
    if c < a:
        t = a
        a = c
        c = t
    if c < b:
        t = b
        b = c
        c = t
    print(a, b, c, ' ')


ex1_2_1()
ex1_2_2()
ex1_2_3()
ex1_2_11()
ex1_2_12()
ex1_2_16()
ex1_2_17()
ex1_2_20()
ex1_2_21()
ex1_2_31()
